lib/ClangImporter/ImportType.cpp

```diff
@@ -2227,6 +2227,157 @@ ImportedType ClangImporter::Implementation::importFunctionParamsAndReturnType(
   return {swiftResultTy, importedType.isImplicitlyUnwrapped()};
 }
 
+static ImportTypeKind
+getImportTypeKindForParam(const clang::ParmVarDecl *param) {
+  ImportTypeKind importKind = ImportTypeKind::Parameter;
+  if (param->hasAttr<clang::CFReturnsRetainedAttr>())
+    importKind = ImportTypeKind::CFRetainedOutParameter;
+  else if (param->hasAttr<clang::CFReturnsNotRetainedAttr>())
+    importKind = ImportTypeKind::CFUnretainedOutParameter;
+
+  return importKind;
+}
+
+Optional<swift::Type> ClangImporter::Implementation::importParameterType(
+    const clang::ParmVarDecl *param, OptionalTypeKind optionalityOfParam,
+    ImportTypeKind importKind, bool allowNSUIntegerAsInt,
+    bool isNSDictionarySubscriptGetter, bool paramIsError,
+    bool paramIsCompletionHandler,
+    Optional<unsigned> completionHandlerErrorParamIndex,
+    ArrayRef<GenericTypeParamDecl *> genericParams,
+    llvm::function_ref<void(Diagnostic &&)> addImportDiagnosticFn,
+    bool &isInOut, bool &isParamTypeImplicitlyUnwrapped) {
+  auto paramTy = param->getType();
+
+  // Import the parameter type into Swift.
+  Type swiftParamTy;
+
+  // Sometimes we import unavailable typedefs as enums. If that's the case,
+  // use the enum, not the typedef here.
+  if (auto typedefType = dyn_cast<clang::TypedefType>(paramTy.getTypePtr())) {
+    if (isUnavailableInSwift(typedefType->getDecl())) {
+      if (auto clangEnum =
+              findAnonymousEnumForTypedef(SwiftContext, typedefType)) {
+        // If this fails, it means that we need a stronger predicate for
+        // determining the relationship between an enum and typedef.
+        assert(clangEnum.getValue()
+                   ->getIntegerType()
+                   ->getCanonicalTypeInternal() ==
+               typedefType->getCanonicalTypeInternal());
+        if (auto swiftEnum = importDecl(*clangEnum, CurrentVersion)) {
+          swiftParamTy = cast<NominalTypeDecl>(swiftEnum)->getDeclaredType();
+        }
+      }
+    }
+  } else if (isa<clang::PointerType>(paramTy) &&
+             isa<clang::TemplateTypeParmType>(paramTy->getPointeeType())) {
+    auto pointeeType = paramTy->getPointeeType();
+    auto templateParamType = cast<clang::TemplateTypeParmType>(pointeeType);
+    PointerTypeKind pointerKind = pointeeType.getQualifiers().hasConst()
+                                      ? PTK_UnsafePointer
+                                      : PTK_UnsafeMutablePointer;
+    auto attrs = getImportTypeAttrs(param, /*isParam=*/true);
```

Comment:
Could these 3 `auto attrs = getImportTypeAttrs(param, /*isParam=*/true);` be hoisted to the top of the function?

Reply:
Two reasons why I moved this into the branches (vs being on top like in the previous version):

- `attrs` was only used inside the branches and inside the branch that does `!swiftParamTy` (Line 2324). In this branch (but not in the others), there's a check for `!swiftParamTy` that returns early, so the Line 2324 will not be hit. Some of these branches do not even use `attrs`, so it was doing extra work for no reason.
- When I saw the differences in `attrs` between the funcion and the method parameters (Line 2329) I notices the `sendableByDefault`, which exist in the method parameters, but not in the function parameters. If I hoist with that value, there will be a check for `sendableByDefault` in the function parameters path which does not make a lot of sense.
- I can also hoist to where it was before, and then rewrite around Line 2329. That would do extra work up front and reuses the variable, which I normally prefer not to do (I prefer variables that do not change values, if possible, but that's my personal style that helps me think about the code).

If nobody thinks that having the `sendableByDefault` in the hoisted version is going to be confusing (or change the behaviour of the code when we do not want it), I can use that version. My order of preference will be the current version I wrote, the one with `sendableByDefault`, and finally the one with the overwritten variable.